functions-continued/variable-scope.md

```diff
@@ -0,0 +1,396 @@
+
+When we want to reliably call a function multiple times, knowing that local variables "leave" after each function call, and "restart" in each function call, creates more reliable code.
+
+In this example, follow these steps in order:
+
+1. Read the code and comments, and trace the values of the local variable after each function call
+2. Predict what the code will print out
+3. Run the code
+4. Compare your predictions against what you actually see
+5. Write down your conclusions about how local variables "leave" at the end of each function call, and get re-initialized at the beginning of each function call.
+
+```python
+def display_breakfast(breakfast_of_champions):
+    # The parameter breakfast_of_champions exists in this function!
+    print(f"This function call has the parameter breakfast_of_champions with the value {breakfast_of_champions}")
+
+    # The local variable breakfast_message exists in this function!
+    breakfast_message = f"My breakfast today is {breakfast_of_champions}"
+    print(f"{breakfast_message}")
+
+    # After the return statement, the function will end, and breakfast_of_champions AND breakfast_message will no longer exist...
+    return breakfast_message
+
+display_breakfast("waffles and syrup")
+# print(f"Uncommenting this line and trying to access breakfast_of_champions will create a NameError: {breakfast_of_champions}")
+# print(f"Uncommenting this line and trying to access breakfast_message will create a NameError: {breakfast_message}")
+display_breakfast("rice and soup")
+display_breakfast("coffee and toast")
+# print(f"Uncommenting this line and trying to access breakfast_of_champions will create a NameError: {breakfast_of_champions}")
+# print(f"Uncommenting this line and trying to access breakfast_message will create a NameError: {breakfast_message}")
+```
+
+At first, it may be frustrating that we can't access `breakfast_of_champions` or `breakfast_message` outside of the function.
+
+On the bright side, it makes our code extremely consistent and predictable: Python will **never** think `breakfast_of_champions` has any value, except when we pass a value into the `display_breakfast` function call. Programming and using variables to know what values they **aren't** is just as valuable as knowing what values they are.
+
+### Our Strategies Will Evolve!
+
+As we learn about more concepts, such as _iteration_ and _classes_, we'll see that local variables have special scope within those kinds of blocks.
+
+It will be crucial to understand the local variable scope for those cases, too, and to rely on them.
+
+Looking ahead: When we learn about _iteration_ in programming, most syntaxes for iteration create their own variable scope.
```

Comment:
Again, this is inaccurate.  There is no special variable scope created. The variable created for iteration has the same  scope as any other variable created in the same context, and lives beyond the end of the iteration.